### PR TITLE
bluestore: Elastic Shared Blobs 2 - move BufferSpace

### PR DESCRIPTION
**Pull request by aclamk**
This is part 2/4 of ESB work.
It executes move of BufferSpace from SharedBlob to Blob.
It is necessary for elastic shared blob logic, but cannot be turn on / off in runtime.
#51439 
#51440 
#51441 
#51442

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -1788,8 +1818,11 @@ class BlueStore : public ObjectStore,
 #endif
     
     std::set<SharedBlobRef> shared_blobs;  ///< these need to be updated/written
+#ifdef WITH_ESB
     std::set<BlobRef> blobs_written; ///< update these on io completion
-
+#else
+    std::set<SharedBlobRef> shared_blobs_written; ///< update these on io completion
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:1824]: nit: we might have a bit less ifdefs if the name doesn't change

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -688,7 +715,10 @@ class BlueStore : public ObjectStore,
       if (--nref == 0)
 	delete this;
     }
+
+#ifdef WITH_ESB
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:719]: nit: we might wrap ~Blob's implementation with ifdef while leaving its declaration for both ESB/non-ESB intact. Just to have fewer ifdefs which generally speaking I don't like...

**Review by ifed01 — APPROVED** [review]

**Comment by ifed01** [comment]
jenkins test make check

**Comment by aclamk** [comment]
@ifed01 Got rid of #ifdef WITH_ESB, last 3 commits are changed.

**Review by ifed01 — APPROVED** [review]

**Comment by rzarzynski** [comment]
Merged as a part of https://github.com/ceph/ceph/pull/53178.

**Pull request closed without merging**